runtime/ISO_Fortran_binding.cc

```diff
@@ -196,17 +213,151 @@ int CFI_is_contiguous(const CFI_cdesc_t *descriptor) {
+    return CFI_INVALID_DESCRIPTOR;
+  }
+  if (source->rank == 0) {
+    return CFI_INVALID_RANK;
+  }
+  if (IsAssumedSize(source) && upper_bounds == nullptr) {
+    return CFI_INVALID_DESCRIPTOR;
+  }
+  if ((result->type != source->type) ||
+      (result->elem_len != source->elem_len)) {
+    return CFI_INVALID_DESCRIPTOR;
+  }
+  if (result->attribute == CFI_attribute_allocatable) {
+    return CFI_INVALID_ATTRIBUTE;
+  }
+  if (source->base_addr == nullptr) {
+    return CFI_ERROR_BASE_ADDR_NULL;
+  }
+
+  char *shiftedBaseAddr{static_cast<char *>(source->base_addr)};
+  bool isZeroSized{false};
+  for (int j{0}; j < source->rank; ++j) {
+    const CFI_dim_t &dim{source->dim[j]};
+    const CFI_index_t srcLB{dim.lower_bound};
+    const CFI_index_t srcUB{srcLB + dim.extent - 1};
+    const CFI_index_t lb{lower_bounds != nullptr ? lower_bounds[j] : srcLB};
+    const CFI_index_t ub{upper_bounds != nullptr ? upper_bounds[j] : srcUB};
+    const CFI_index_t stride{strides != nullptr ? strides[j] : 1};
+
+    if (stride == 0 && lb != ub) {
+      return CFI_ERROR_OUT_OF_BOUNDS;
+    }
+    if ((lb <= ub && stride >= 0) || (lb >= ub && stride < 0)) {
+      if ((lb < srcLB) || (lb > srcUB) || (ub < srcLB) || (ub > srcUB)) {
+        return CFI_ERROR_OUT_OF_BOUNDS;
+      }
+      shiftedBaseAddr += (lb - srcLB) * dim.sm;
+      extent[j] = stride != 0 ? 1 + (ub - lb) / stride : 1;
+    } else {
+      isZeroSized = true;
+      extent[j] = 0;
+    }
+    actualStride[j] = stride;
+    resRank += (stride != 0);
+  }
+  if (resRank != result->rank) {
+    return CFI_INVALID_DESCRIPTOR;
+  }
+
+  // For zero-sized arrays, base_addr is processor-dependent (see 18.5.3).
+  // We keep it on the source base_addr
+  result->base_addr = isZeroSized ? source->base_addr : shiftedBaseAddr;
+  resRank = 0;
+  for (int j{0}; j < source->rank; ++j) {
+    if (actualStride[j] != 0) {
+      result->dim[resRank].lower_bound = 0;
+      result->dim[resRank].extent = extent[j];
+      result->dim[resRank].sm = actualStride[j] * source->dim[j].sm;
+      ++resRank;
+    }
+  }
+  return CFI_SUCCESS;
 }
 
 int CFI_select_part(CFI_cdesc_t *result, const CFI_cdesc_t *source,
     std::size_t displacement, std::size_t elem_len) {
-  return CFI_INVALID_DESCRIPTOR;  // TODO
+  if (result == nullptr || source == nullptr) {
+    return CFI_INVALID_DESCRIPTOR;
+  }
+  if (result->rank != source->rank) {
+    return CFI_INVALID_RANK;
+  }
+  if (result->attribute == CFI_attribute_allocatable) {
+    return CFI_INVALID_ATTRIBUTE;
+  }
+  if (source->base_addr == nullptr) {
+    return CFI_ERROR_BASE_ADDR_NULL;
+  }
+  if (IsAssumedSize(source)) {
+    return CFI_INVALID_DESCRIPTOR;
+  }
+
+  if (!IsCharacterType(result->type)) {
+    elem_len = result->elem_len;
+  }
+  if (displacement + elem_len > source->elem_len) {
+    return CFI_INVALID_ELEM_LEN;
+  }
+
+  result->base_addr = reinterpret_cast<void *>(
+      displacement + reinterpret_cast<char *>(source->base_addr));
```

Comment:
Another static_cast<> opportunity.  Since the base_addr member is declared void*, it might make sense to consider adding some functions to freely cast it to char* (and back to void* for assignments).  The idea being to put functions as a firewall between using the pointer and the pointer's (required but not very amenable) declared type.

Reply:
That is a good idea, there is actually already something doing a firewall over this pointer runtime/descriptor.h. When the pointer is seen as a Descriptor object it is possible to use the member function Element<T>(offset) to get a pointer of type T.
I do not really want to use it here because I would rather stay away from the Descriptor class when implementing ISO required functions and work only with the ISO descriptor (CFI_cdesc_t). In the rest of the rest runtime, I will follow your suggestion and avoid repeating the cast.